src/clp_ffi_py/ir/Query.hpp

```diff
@@ -0,0 +1,200 @@
+     * @param wildcard Wildcard string.
+     * @param case_sensitive Case sensitive indicator.
+     */
+    WildcardQuery(std::string wildcard, bool case_sensitive)
+            : m_wildcard(std::move(wildcard)),
+              m_case_sensitive(case_sensitive){};
+
+    [[nodiscard]] auto get_wildcard() const -> std::string_view { return m_wildcard; }
+
+    [[nodiscard]] auto is_case_sensitive() const -> bool { return m_case_sensitive; }
+
+private:
+    std::string m_wildcard;
+    bool m_case_sensitive;
+};
+
+/**
+ * This class represents a search query, utilized for filtering log events in a
+ * CLP IR stream. The query could include a list of wildcard searches aimed at
+ * identifying certain log messages, and a timestamp range with a lower and
+ * upper bound. This class provides an interface to set up a search query, as
+ * well as methods to validate whether the query can be matched by a log event.
+ * Note that an empty wildcard query list will match any log within the range.
+ */
+class Query {
+public:
+    static constexpr ffi::epoch_time_ms_t const cTimestampMin{0};
+    static constexpr ffi::epoch_time_ms_t const cTimestampMax{
+            std::numeric_limits<ffi::epoch_time_ms_t>::max()};
+
+    /**
+     * Ideally, when searching an IR stream with a query, the search terminates
+     * once the timestamp exceeds the search upper bound. However, the timestamp
+     * might not be monotonically increasing in a CLP IR stream. It can be
+     * locally disordered due to thread contention. To safely stop searching, we
+     * need to ensure that the current timestamp has exceeded the search's upper
+     * bound by a reasonable margin. This margin defaults to the following
+     * constant. During the query initialization, it will be applied to the
+     * upper bound timestamp to generate a timestamp which indicates it is safe
+     * to terminate the search.
+     */
+    static constexpr ffi::epoch_time_ms_t const cDefaultSearchTerminationMargin{
+            static_cast<ffi::epoch_time_ms_t>(60 * 1000)};
+
+    /**
+     * Constructs an empty query object that will match all logs. The wildcard
+     * list is empty and the timestamp range is set to include all the valid
+     * Unix epoch timestamps.
+     */
+    explicit Query()
+            : m_lower_bound_ts{cTimestampMin},
+              m_upper_bound_ts{cTimestampMax},
+              m_search_termination_ts{cTimestampMax} {};
+
+    /**
+     * Constructs a new query object with the given timestamp range with an
+     * empty wildcard list.
+     * @param search_time_lower_bound Start of search time range (inclusive).
+     * @param search_time_upper_bound End of search time range (inclusive).
+     * @param search_termination_margin The margin used to determine the search
+     * termination timestamp.
+     */
+    explicit Query(
+            ffi::epoch_time_ms_t search_time_lower_bound,
+            ffi::epoch_time_ms_t search_time_upper_bound,
```

Comment:
How about `time_range_begin` and `time_range_end`?

Reply:
I would stick to `lower/upper bound`